server/src/main/java/org/elasticsearch/transport/ConnectionManager.java

```diff
@@ -122,40 +117,34 @@ public void connectToNode(DiscoveryNode node, ConnectionProfile connectionProfil
             return;
         }
 
-        try (Releasable lock = connectionLock.acquire(node.getId())) {
-            Transport.Connection connection = connectedNodes.get(node);
-            if (connection != null) {
-                assert connectingNodes.containsKey(node) == false;
-                lock.close();
-                connectingRefCounter.decRef();
-                listener.onResponse(null);
-                return;
-            }
+        if (pendingConnections.containsKey(node)) {
+            connectingRefCounter.decRef();
+            listener.onResponse(null);
+            return;
+        }
 
-            final List<ActionListener<Void>> connectionListeners = connectingNodes.computeIfAbsent(node, n -> new ArrayList<>());
-            connectionListeners.add(listener);
-            if (connectionListeners.size() > 1) {
-                // wait on previous entry to complete connection attempt
-                connectingRefCounter.decRef();
-                return;
-            }
+        final CompletableContext<Void> currentAttemptContext = new CompletableContext<>();
```

Comment:
Should we use `ListenableFuture`? `CompletableContext` is not safe I think (if the listener throws an error, then that gets swallowed I think instead of being bubbled up to the uncaught exception handler).